sklearn/tree/_splitter.pyx

```diff
@@ -550,199 +507,296 @@ cdef void heapsort(DTYPE_t* Xf, SIZE_t* samples, SIZE_t n) nogil:
+            continue
+
+        # f_j in the interval [n_known_constants, f_i - n_found_constants[
+        f_j += n_found_constants
+        # f_j in the interval [n_total_constants, f_i[
+
+        current.feature = features[f_j]
+
+        # Find min, max
+        data_splitter.find_min_max(current.feature, &min_feature_value, &max_feature_value)
+
+        if max_feature_value <= min_feature_value + FEATURE_THRESHOLD:
+            features[f_j], features[n_total_constants] = features[n_total_constants], current.feature
+
+            n_found_constants += 1
+            n_total_constants += 1
+            continue
+
+        f_i -= 1
+        features[f_i], features[f_j] = features[f_j], features[f_i]
+
+        # Draw a random threshold
+        current.threshold = rand_uniform(min_feature_value,
+                                         max_feature_value,
+                                         random_state)
+
+        if current.threshold == max_feature_value:
+            current.threshold = min_feature_value
+
+        # Partition
+        current.pos = data_splitter.parition_samples(current.threshold)
+
+        # Reject if min_samples_leaf is not guaranteed
+        if (((current.pos - start) < min_samples_leaf) or
+                ((end - current.pos) < min_samples_leaf)):
+            continue
+
+        # Evaluate split
+        criterion.reset()
+        criterion.update(current.pos)
+
+        # Reject if min_weight_leaf is not satisfied
+        if ((criterion.weighted_n_left < min_weight_leaf) or
+                (criterion.weighted_n_right < min_weight_leaf)):
+            continue
+
+        current_proxy_improvement = criterion.proxy_impurity_improvement()
+
+        if current_proxy_improvement > best_proxy_improvement:
+            best_proxy_improvement = current_proxy_improvement
+            best = current  # copy
+
+    # Reorganize into samples[start:best.pos] + samples[best.pos:end]
+    if best.pos < end:
+        if current.feature != best.feature:
+            data_splitter.parition_samples_best(best.pos, best.threshold, best.feature)
+
+        criterion.reset()
+        criterion.update(best.pos)
+        criterion.children_impurity(&best.impurity_left,
+                                    &best.impurity_right)
+        best.improvement = criterion.impurity_improvement(
+            impurity, best.impurity_left, best.impurity_right)
+
+    # Respect invariant for constant features: the original order of
+    # element in features[:n_known_constants] must be preserved for sibling
+    # and child nodes
+    memcpy(&features[0], &constant_features[0], sizeof(SIZE_t) * n_known_constants)
+
+    # Copy newly found constant features
+    memcpy(&constant_features[n_known_constants],
+           &features[n_known_constants],
+           sizeof(SIZE_t) * n_found_constants)
+
+    # Return values
+    split[0] = best
+    n_constant_features[0] = n_total_constants
+    return 0
+
+@final
```

Comment:
Related to https://github.com/scikit-learn/scikit-learn/pull/25306/files#r1062801209. Would it make sense to have a base class for "Dense" and "Sparse" splitting that is not `final`? 

And the fused type `DataSplitter` can come from the base classes rather than the final implementations?

Reply:
I tried using a base class and it runs ~10% slowly compared to `main`. I think the overhead comes from many virtual table lookups. `node_split_*` calls into `data_splitter._` a lot and `node_split_*` is called a lot when building the trees.

Reply:
BTW there can be a parent class `DataSplitter` as part of the fused type to allow subclasses,  but the `DenseSplitter` and `SparseSplitter` needs to be `final`. The `final` is required because it avoids any vtable lookups leading to a ~10% runtime regression.

I prefer this PR to focus on the refactor and not add new third party extension points.